### ISSUE TITLE
gprestore interrupt when recover index

### ISSUE THREAD

**Issue by cobolbaby** [greenplum-db/gpbackup#423]
When performing index recovery, the error will appear. The related logs and SQL are as follows:

- gprestore log

```
20201018:14:06:34 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Restore Command: [gprestore --backup-dir /gpbackup/20201016185432 --timestamp 20201016185716 --metadata-only --create-db --jobs 8]
20201018:14:06:34 gprestore:gpadmin:gp6mdw:024511-[INFO]:-Restore Key = 20201016185716
20201018:14:06:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Metadata will be restored from /gpbackup/20201016185432/gpseg-1/backups/20201016/20201016185716/gpbackup_20201016185716_metadata.sql
20201018:14:06:35 gprestore:gpadmin:gp6mdw:024511-[INFO]:-Creating database
20201018:14:06:37 gprestore:gpadmin:gp6mdw:024511-[INFO]:-Database creation complete for: "F6_BDC"
20201018:14:06:38 gprestore:gpadmin:gp6mdw:024511-[WARNING]:-This backup set was taken on a version of Greenplum prior to 6.x. This restore will use the legacy hash operators when loading data.
20201018:14:06:38 gprestore:gpadmin:gp6mdw:024511-[WARNING]:-To use the new Greenplum 6.x default hash operators, these tables will need to be redistributed.
20201018:14:06:38 gprestore:gpadmin:gp6mdw:024511-[WARNING]:-For more information, refer to the migration guide located as https://docs.greenplum.org/latest/install_guide/migrate.html.
20201018:14:06:38 gprestore:gpadmin:gp6mdw:024511-[INFO]:-Restoring pre-data metadata
20201018:14:07:02 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  10% (199/1981)
20201018:14:07:27 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  20% (397/1981)
20201018:14:07:59 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  30% (595/1981)
20201018:14:08:24 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  40% (793/1981)
20201018:14:08:56 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  50% (991/1981)
20201018:14:12:40 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  60% (1189/1981)
20201018:14:14:33 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  70% (1387/1981)
20201018:14:16:08 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  80% (1585/1981)
20201018:14:17:03 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  90% (1783/1981)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Pre-data objects restored:  100% (1981/1981)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[INFO]:-Pre-data metadata restore complete
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[INFO]:-Restoring post-data metadata
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Error encountered when executing statement: _ssd01;

CREATE INDEX fact_shipping_info_wanglin2_mcbsno_pack_date ON dw.fact_shipping_info_wanglin2 USING btree (mcbsno, pack Error was: ERROR: syntax error at or near "_ssd01" (SQLSTATE 42601)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Error encountered when executing statement: _ssd01;

CREATE INDEX fact_shipping_info_rma_mcbsno_rma_no ON dw.fact_shipping_info_rma USING btree (mcbsno, r Error was: ERROR: syntax error at or near "_ssd01" (SQLSTATE 42601)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Error encountered when executing statement: (id);


CREATE INDEX fact_shipping_info_lr_combine_mcbsno_rma_no ON dw.fact_shipping_info_lr_combine USING btree (mcbsno, r Error was: ERROR: syntax error at or near "id" (SQLSTATE 42601)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Error encountered when executing statement: (cdt);

CREATE INDEX idx_dw_customer_datasource_customer ON dw.customer_datasource USING btree (cus Error was: ERROR: syntax error at or near "cdt" (SQLSTATE 42601)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Error encountered when executing statement: _ssd01;

CREATE INDEX fact_shipping_info_lr_not_combine_mcbsno_rma_no ON dw.fact_shipping_info_lr_not_combine USING btree (mcbsno, r Error was: ERROR: syntax error at or near "_ssd01" (SQLSTATE 42601)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Error encountered when executing statement: _ssd01;

CREATE INDEX fact_shipping_info_pca_combine_mcbsno_rma_no ON dw.fact_shipping_info_pca_combine USING btree (mcbsno, r Error was: ERROR: syntax error at or near "_ssd01" (SQLSTATE 42601)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Error encountered when executing statement: _date);

CREATE INDEX i_cdt ON fis.pca_pca_lot USING btree Error was: ERROR: syntax error at or near "_date" (SQLSTATE 42601)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Error encountered when executing statement: _ssd01;

CREATE INDEX idx_dw_dim_cpu_dn_dn ON dw.dim_cpu_dn USING btre Error was: ERROR: syntax error at or near "_ssd01" (SQLSTATE 42601)
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[CRITICAL]:-ERROR: syntax error at or near "_ssd01" (SQLSTATE 42601)
github.com/greenplum-db/gpbackup/restore.ExecuteStatements
        /tmp/build/3e49593f/go/src/github.com/greenplum-db/gpbackup/restore/parallel.go:79
github.com/greenplum-db/gpbackup/restore.ExecuteRestoreMetadataStatements
        /tmp/build/3e49593f/go/src/github.com/greenplum-db/gpbackup/restore/wrappers.go:280
github.com/greenplum-db/gpbackup/restore.restorePostdata
        /tmp/build/3e49593f/go/src/github.com/greenplum-db/gpbackup/restore/restore.go:348
github.com/greenplum-db/gpbackup/restore.DoRestore
        /tmp/build/3e49593f/go/src/github.com/greenplum-db/gpbackup/restore/restore.go:150
main.main.func1
        /tmp/build/3e49593f/go/src/github.com/greenplum-db/gpbackup/gprestore.go:23
github.com/spf13/cobra.(*Command).execute
        /tmp/build/3e49593f/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830
github.com/spf13/cobra.(*Command).ExecuteC
        /tmp/build/3e49593f/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914
github.com/spf13/cobra.(*Command).Execute
        /tmp/build/3e49593f/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main
        /tmp/build/3e49593f/go/src/github.com/greenplum-db/gpbackup/gprestore.go:27
runtime.main
        /usr/local/go/src/runtime/proc.go:203
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1357
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[INFO]:-Found neither /usr/local/greenplum-db-6.11.2/bin/gp_email_contacts.yaml nor /home/gpadmin/gp_email_contacts.yaml
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[INFO]:-Email containing gprestore report /gpbackup/20201016185432/gpseg-1/backups/20201016/20201016185716/gprestore_20201016185716_20201018140634_report will not be sent
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Beginning cleanup
20201018:14:17:35 gprestore:gpadmin:gp6mdw:024511-[DEBUG]:-Cleanup complete
```

- sql 

```
CREATE INDEX fact_shipping_info_lr_combine_mcbsno_rma_no ON dw.fact_shipping_info_lr_combine USING btree (mcbsno, rma_no);
ALTER INDEX dw.fact_shipping_info_lr_combine_mcbsno_rma_no SET TABLESPACE tbs_ssd01;

CREATE INDEX fact_shipping_info_lr_not_combine_mcbsno_rma_no ON dw.fact_shipping_info_lr_not_combine USING btree (mcbsno, rma_no);
ALTER INDEX dw.fact_shipping_info_lr_not_combine_mcbsno_rma_no SET TABLESPACE tbs_ssd01;

CREATE INDEX fact_shipping_info_pca_combine_mcbsno_rma_no ON dw.fact_shipping_info_pca_combine USING btree (mcbsno, rma_no);
ALTER INDEX dw.fact_shipping_info_pca_combine_mcbsno_rma_no SET TABLESPACE tbs_ssd01;

CREATE INDEX fact_shipping_info_rma_mcbsno_rma_no ON dw.fact_shipping_info_rma USING btree (mcbsno, rma_no);
ALTER INDEX dw.fact_shipping_info_rma_mcbsno_rma_no SET TABLESPACE tbs_ssd01;

CREATE INDEX fact_shipping_info_wanglin2_mcbsno_pack_date ON dw.fact_shipping_info_wanglin2 USING btree (mcbsno, pack_date);

CREATE INDEX i_cdt ON fis.pca_pca_lot USING btree (cdt);

CREATE INDEX idx_dw_customer_datasource_customer ON dw.customer_datasource USING btree (customer);
ALTER INDEX dw.idx_dw_customer_datasource_customer SET TABLESPACE tbs_ssd01;

CREATE INDEX idx_dw_dim_cpu_dn_dn ON dw.dim_cpu_dn USING btree (dn);
ALTER INDEX dw.idx_dw_dim_cpu_dn_dn SET TABLESPACE tbs_ssd01;
```

**Comment by jimmyyih** [comment]
Hi @cobolbaby, thanks for reporting an issue.  I was unable to reproduce your issue by doing the following:
1. On GPDB 5 cluster, create a table and btree index.  Alter btree index to be on a user-defined tablespace (requires filespace because of GPDB 5).
2. Create a backup using gpbackup and --backup-dir flag.
3. On GPDB 6 cluster, manually create the user-defined tablespace (tablespaces are different in GPDB 6 so need to create it manually with new syntax).
4. On GPDB 6 cluster, run gprestore with the --timestamp and --backup-dir from the backup.

My metadata SQL file had index info that looked like this (similar to what you posted):
```
CREATE INDEX fooidx ON fooschema.foobar USING btree (b, c);
ALTER INDEX fooschema.fooidx SET TABLESPACE mytablespace;
```

Looking at your gprestore log output, it looks like the metadata SQL file could have been modified and is no longer matching with the metadata TOC file (table of contents file that tracks byte offsets of each SQL command in the metadata SQL file).  Did you or someone else happen to modify the metadata SQL file (check if it still has 444 file permissions maybe)?  Also, could you check the GPDB master segment logs (`$MASTER_DATA_DIRECTORY/pg_log/`) and copy/paste the full error messages related to your gprestore errors to get the complete error message output?

**Comment by cobolbaby** [comment]
Yes, I did modify the metadata SQL file to fix the bug #424

**Comment by jimmyyih** [comment]
> Can I manually modify the TOC file?

Technically, you could modify the metadata TOC file but it is strongly discouraged.  The metadata TOC file is essentially an index for the metadata SQL file and you would have to modify many entries to correct the start/end byte offsets to match your modified metadata SQL file.

I would suggest reverting your changes to the metadata SQL file if you made a copy and use the workarounds I provided in issue #424.
